File: Tutorials/regular-expressions/assets/regex example.py
#############################################
#=============Read in Libraries=============#
# Read in the necessary libraries.          #
#############################################

#import re
import regex


#############################################
#===========Regex Matched Object============#
# The following examples use the functions  #
# match() and search().                     #
#############################################

#===============================
# First example, using match()
#===============================
reg_string1 = "To catch a cat requires skill not caterpillar like slowness or cataract like myopia"
reg_string2 = "Cats exude cataract like myopia"
pattern1 = r"[Cc]at."

# The function match() only compares the 
# first word in a string
if regex.match(pattern1, reg_string1):
    print("You found a match!")
else: print("Bummer. No match.")

if regex.match(pattern1, reg_string2):
    print("You found a match!")
else: print("Bummer. No match.")


#=================================
# Second example, using search()
#=================================
if regex.search(pattern1, reg_string1):
    print("You found a match!")
else: print("Bummer. No match.")

if regex.search(pattern1, reg_string2):
    print("You found a match!")
else: print("Bummer. No match.")


#===========================================
# Comparing functions match() and search()
#===========================================
matches1 = regex.match(pattern1, reg_string1)
matches2 = regex.match(pattern1, reg_string2)
matches3 = regex.search(pattern1, reg_string1)
matches4 = regex.search(pattern1, reg_string2)
matches5 = regex.search("[Dd][Oo][Gg]", reg_string1)
matches6 = regex.search("[Dd][Oo][Gg]", reg_string2)

matches1
matches2
matches3
matches4
matches5
matches6

#===================================================
# Appending group() at the end of the match object
#===================================================
matches2.group()
matches3.group()
matches4.group()


#==========================================
# Other functions: span(), start(), end()
#==========================================
matches2.span()
matches2.start()
matches2.end()


#============================================
# Using group() to extract grouped patterns
#============================================
reg_email = r"([a-z\.-]+)@([a-z\.-]+)\.(com|edu)"
pattern_email1 = "bryan.hammer@okstate.edu"

matches7 = regex.search(reg_email, pattern_email1)
matches7.group()
matches7.group(1)
matches7.group(2)
matches7.group(3)


#========================================
# Use the function findall() to extract 
# all possible matches in a string
#========================================
reg_string3 = "To catch a cat requires skill not caterpillar like slowness or cataract like myopia"
pattern2 = r"(cat){1}([a-z]*)"

matches8 = regex.findall(pattern2,reg_string3)

matches8

# in the loop, cats is arbitrary; you could use
# asdf instead of cats
for cats in matches8: 
    print(cats)

len(matches8)


#====================================
# Pulling email addresses from html
#====================================
reg_string4 = '<p><a href="mailto:bryan.hammer@okstate.edu?subject=">bryan.hammer@okstate.edu</a></p><p><a href="mailto:ramesh.sharda@okstate.edu?subject=">ramesh.sharda@okstate.edu</a></p>'
reg_ptags = r"<p><a href=\"mailto:([a-z\.-]+)@([a-z\.-]+)\.(com|edu)\?subject=\">([a-z\.-]+)@([a-z\.-]+)\.(com|edu)</a></p>"

matches9 = regex.findall(reg_ptags,reg_string4)

for emaillist in matches9:
    print(emaillist)

# Another possibility
reg_ptags2 = r"<p><a href=\"mailto:(?P<username>[a-z\.-]+)@(?P<business>[a-z\.-]+)\.(?P<entitytype>com|edu)\?subject=\">([a-z\.-]+)@([a-z\.-]+)\.(com|edu)</a></p>"

for i in regex.finditer(reg_ptags2,reg_string4):
    print(i.group('username'), i.group('business'), i.group('entitytype'))


#############################################
#=========Regex Splitting Strings===========#
# The following examples use the function   #
# split() to create substrings.             #
#############################################

reg_string5 = reg_string1.split()
reg_string5

# Illustration of whitespace as delimiter
# Two additional spaces added around "cat"
reg_string6 = "To catch a  cat  requires skill not caterpillar like slowness or cataract like myopia"
reg_string7 = reg_string6.split(" ")
reg_string7

reg_string6 = "To catch a  cat  requires skill not caterpillar like slowness or cataract like myopia"
reg_string7 = reg_string6.split(None)
reg_string7

File: Tutorials/regular-expressions/assets/regex example.R
#####################################################
#============Setup the Working Directory============#
# Set the working directory to the project folder by#
# running the appropriate code below.               #
#####################################################

wd_home = "C:\\Users\\bryan\\OneDrive - Oklahoma State University\\Teaching\\MSIS 5193 4623 - Programming Data 1\\Data"
setwd(wd_home)

wd_office = "C:\\Users\\bryan\\OneDrive - Oklahoma A and M System\\Teaching\\MSIS 5193 4623 - Programming Data 1\\Data"
setwd(wd_office)


#############################################
#==============Regex with grep==============#
# The following examples use the basic      #
# function grep.                            #
#############################################

#==========================================================
# First example, using quantifiers for matching patterns
#==========================================================
reg_string1 = c('x', 'xy', 'xyz', 'xyyz', 'xyyyz', 'xyyyyz', 'xz')

grep("xy*z", reg_string1, value = TRUE)
grep("xy+z", reg_string1, value = TRUE)
grep("yy+", reg_string1, value = TRUE)
grep("xy?z", reg_string1, value = TRUE)
grep("yy?", reg_string1, value = TRUE)
grep("yy{1}", reg_string1, value = TRUE)
grep("xy{2}z", reg_string1, value = TRUE)
grep("xy{2,}z", reg_string1, value = TRUE)
grep("xy{2,3}z", reg_string1, value = TRUE)
grep("y{2,3}", reg_string1, value = TRUE)

#==================================
# Using anchors to match patterns
#==================================
reg_string2 = c('abcd', 'cdab', 'cabd', 'c abd')

grep("ab", reg_string2, value = TRUE)
grep("^ab", reg_string2, value = TRUE)
grep("bd$", reg_string2, value = TRUE)
grep("\\bab", reg_string2, value = TRUE)

#=============================
# Using the . as a wild card
#=============================
reg_string3 = c('catch', 'cater', 'ducat', 'scatter', 'locate')

grep(".cat", reg_string3, perl = TRUE, value = TRUE)
grep("cat.", reg_string3, perl = TRUE, value = TRUE)
grep(".cat.", reg_string3, perl = TRUE, value = TRUE)


#==========================================
# Real-world example, searching for "cat"
#==========================================
reg_string4 = c('To', 'catch', 'a', 'cat', 'requires', 'skill', 'not', 'caterpillar', 'like', 'slowness', 'or', 'cataract', 'like', 'myopia', 'asdcat')

grep("cat", reg_string4, perl = TRUE, value = FALSE)
grep("cat", reg_string4, perl = TRUE, value = TRUE)

grep("cat$", reg_string4, perl = TRUE, value = TRUE)

#===========================
# More real-world examples
#===========================
reg_string5 = c('catch', 'cat', 'caterpillar', 'cataract', 'asdcat', 'xycatzz', 'abc cat xyz')

grep("^cat", reg_string5, perl = TRUE, value = TRUE)
grep("cat$", reg_string5, perl = TRUE, value = TRUE)
grep("^cat$", reg_string5, perl = TRUE, value = TRUE)
grep("\\bcat\\b", reg_string5, perl = TRUE, value = TRUE)


#############################################
#=============Tidyverse Stringr=============#
# Use the functions contained in the library#
# stringr to identify, match, detect, count #
# regular expression patterns.              #
#############################################

#install.packages("stringr")
library(stringr)

#=========================================
# Using str_detect() to identify matches
#=========================================
reg_string6 = c('To', 'catch', 'a', 'cat', 'requires', 'skill', 'not', 'caterpillar', 'like', 'slowness', 'or', 'cataract', 'like', 'myopia')
str_detect(reg_string6, "cat")
sum(str_detect(reg_string6, "cat"))

str_subset(reg_string6, "cat")

#================================
# Words that do not contain cat
#================================
(no_cat = !str_detect(reg_string6,"cat"))

#==============
# str_count()
#==============
reg_string7 = c('catch', 'cat', 'caterpillar', 'catcatcat', 'catcat', 'xycatzz', 'abc cat xyz')
str_count(reg_string7, "cat")

#================================
# str_subset() vs str_extract()
#================================
reg_string8 = c('To catch a cat requires', 'skill', 'not caterpillar-like slowness or', 'cataract', 'like myopia')
str_subset(reg_string8, "cat")

matched_strings1 = str_subset(reg_string8, "cat")
matches1 = str_extract(matched_strings1, "cat")
matches1

# Extracting all the values from a string, not just the first
matches_more = str_extract_all(matched_strings1, "cat")
matches_more

matches_more = str_extract_all(matched_strings1, "cat", simplify = TRUE)
matches_more

# Another example
length(fruit)
head(fruit)

fruit_types = c('berry', 'fruit')
fruit_regex = str_c(fruit_types, collapse = "|")
fruit_regex

matched_strings2 = str_subset(fruit, fruit_regex)
matched_strings2

matches2 = str_extract(matched_strings2, fruit_regex)
matches2

#=================================
# Using the function str_match()
#=================================
str_match(reg_string6, "(.)[aeiou](.)")

# A more complex example
noun = "(a|the) ([^ ]+)"
matched_strings3 = str_subset(sentences, noun)
matches3 = str_extract(matched_strings3, noun)
head(matches3)
matches4 = str_match(matched_strings3, noun)
head(matches4)
